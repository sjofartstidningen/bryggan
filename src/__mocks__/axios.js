/* eslint-disable no-underscore-dangle */
function axios() {
  return Promise.resolve(axios.__return);
}

axios.create = function create() {
  return axios;
};

axios.isCancel = () => false;

axios.defaults = {
  adapter: () => null,
};

axios.__return = {
  data: {
    hello: 'world',
  },
};

axios.__setReturn = ret => {
  axios.__return = ret;
};

export class CancelToken {
  static source() {
    return {
      token: 'token',
      cancel: () => 'cancel',
    };
  }
}

axios.CancelToken = CancelToken;

export default axios;
