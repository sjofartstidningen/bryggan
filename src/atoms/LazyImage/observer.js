// @flow
import mitt from 'mitt';

const emitter = mitt();
const events = {
  IN_VIEW: 'in-view',
  OUT_OF_VIEW: 'out-of-view',
};

function createObserver() {
  let observer: ?IntersectionObserver;

  return function getObserver(): IntersectionObserver {
    if (observer) return observer;

    observer = new IntersectionObserver(
      entries => {
        entries.forEach(entry => {
          if (entry.isIntersecting) emitter.emit(events.IN_VIEW, entry);
          else emitter.emit(events.OUT_OF_VIEW, entry);
        });
      },
      { root: null, rootMargin: '50px', thresholds: 0 },
    );

    return observer;
  };
}

export default createObserver();
export { emitter, events };
