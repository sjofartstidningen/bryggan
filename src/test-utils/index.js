/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter } from 'react-router-dom';
import { theme } from '../styles';

const WithContexts = ({ children }) => (
  <BrowserRouter>
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  </BrowserRouter>
);

function setupIntersectionObserver() {
  const original = window.IntersectionObserver;

  class IntersectionObserver {
    constructor(callback) {
      this.callback = callback;
    }

    entries = [];

    observe = el => {
      this.entries.push(el);
    };

    unobserve = el => {
      this.entries = this.entries.filter(x => x !== el);
    };

    simulateIntersectionEvent = (data = {}) => {
      const els = this.entries.map(entry => ({
        target: entry,
        ...data,
      }));

      this.callback(els);
    };
  }

  window.IntersectionObserver = IntersectionObserver;

  return () => {
    window.IntersectionObserver = original;
  };
}

const sleep = (ms = 0) =>
  new Promise(resolve => {
    window.setTimeout(resolve, ms);
  });

export { WithContexts, setupIntersectionObserver, sleep };
