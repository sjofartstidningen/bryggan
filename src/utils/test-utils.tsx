import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/theme';
import { AuthProvider } from '../hooks/useAuth';

const Providers: React.FC = ({ children }) => {
  return (
    <React.Suspense fallback={<p>Loading</p>}>
      <ThemeProvider theme={theme}>
        <AuthProvider>{children}</AuthProvider>
      </ThemeProvider>
    </React.Suspense>
  );
};

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, {
    wrapper: ({ children }) => <Providers>{children}</Providers>,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
