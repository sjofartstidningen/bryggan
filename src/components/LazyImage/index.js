import React, { Component } from 'react';
import PropTypes from 'prop-types';
import axios, { CancelToken } from 'axios';
import { cacheAdapterEnhancer } from 'axios-extensions';
import mitt from 'mitt';

const emitter = mitt();

const http = axios.create({
  headers: {
    'Cache-Control': 'no-cache',
  },
  adapter: cacheAdapterEnhancer(axios.defaults.adapter, true),
});

const events = {
  IN_VIEW: 'in-view',
  OUT_OF_VIEW: 'out-out-view',
};

const observer = new IntersectionObserver(
  entries => {
    entries.forEach(entry => {
      if (entry.isIntersecting) emitter.emit(events.IN_VIEW, entry);
      else emitter.emit(events.OUT_OF_VIEW, entry);
    });
  },
  {
    root: null,
    rootMargin: '100px',
    thresholds: 0,
  },
);

class LazyImage extends Component {
  static propTypes = {
    src: PropTypes.string.isRequired,
    render: PropTypes.func.isRequired,
    onLoad: PropTypes.func,
    onError: PropTypes.func,
  };

  static defaultProps = {
    onLoad: null,
    onError: null,
  };

  state = {
    src: null,
    loaded: false,
    error: null,
  };

  componentDidMount() {
    observer.observe(this.wrapper);
    emitter.on(events.IN_VIEW, this.handleIntersect);
    emitter.on(events.OUT_OF_VIEW, this.handleNonIntersect);
  }

  componentWillUnmount() {
    this.unmountListeners();
  }

  getImage = async ({ cache = true } = {}) => {
    try {
      this.cancelToken = CancelToken.source();
      const res = await http({
        method: 'get',
        url: this.props.src,
        responseType: 'blob',
        cancelToken: this.cancelToken.token,
        cache,
      });

      const src = URL.createObjectURL(res.data);
      this.setState(() => ({ src, loaded: true }));
      this.unmountListeners();

      if (this.props.onLoad != null) this.props.onLoad(res);
    } catch (err) {
      if (!axios.isCancel(err)) this.handleError(err);
    }
  };

  unmountListeners = () => {
    observer.unobserve(this.wrapper);
    emitter.off(events.IN_VIEW, this.handleIntersect);
    emitter.off(events.OUT_OF_VIEW, this.handleNonIntersect);
    if (this.cancelToken) {
      this.cancelToken.cancel('Stop listening for events');
    }
  };

  revokeObjectURL = ({ target }) => URL.revokeObjectURL(target.src);

  handleIntersect = entry => {
    if (entry.target === this.wrapper) this.getImage();
  };

  handleNonIntersect = entry => {
    if (entry.target === this.wrapper && this.cancelToken) {
      this.cancelToken.cancel('Component out of view');
    }
  };

  handleError = error => {
    this.setState(() => ({ error }));
    this.unmountListeners();
    if (this.props.onError != null) this.props.onError(error);
  };

  handleReload = e => {
    e.preventDefault();
    e.stopPropagation();
    this.setState({ loaded: false });
    this.getImage({ cache: false });
  };

  render() {
    const { render } = this.props;
    const { src, loaded, error } = this.state;

    return (
      <div
        ref={ref => {
          this.wrapper = ref;
        }}
      >
        {render({
          src,
          loaded,
          error,
          revokeObjectURL: this.revokeObjectURL,
          reload: this.handleReload,
        })}
      </div>
    );
  }
}

export default LazyImage;
