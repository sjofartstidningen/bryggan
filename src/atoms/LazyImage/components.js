import React from 'react';
import styled, { css } from 'styled-components';
import { hideVisually } from 'polished';
import { Close } from '../../atoms/Icon';

const Container = styled.div.attrs({
  ratio: p => p.ratio || null,
})`
  position: relative;
  width: 100%;
  height: auto;

  ${p =>
    p.ratio &&
    css`
      height: 0;
      padding-bottom: calc(100% * ${p.ratio});
    `};
`;

const HiddenImage = styled.img`
  ${hideVisually()};
`;

const DefaultErrorWrapper = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-size: 1.5rem;
  color: ${({ theme }) => theme.color.error};
`;

function DefaultError() {
  return (
    <DefaultErrorWrapper>
      <Close />
    </DefaultErrorWrapper>
  );
}

export { Container, HiddenImage, DefaultError };
