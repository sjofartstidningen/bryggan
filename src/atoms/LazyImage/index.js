import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Container, Img, HiddenImage, DefaultError } from './components';
import getObserver, { emitter, events } from './observer';
import ProgressBar from '../ProgressBar';

class LazyImage extends PureComponent {
  static propTypes = {
    src: PropTypes.string.isRequired,
    alt: PropTypes.string,
    ratio: PropTypes.number,
    className: PropTypes.string,
    onLoad: PropTypes.func,
    onError: PropTypes.func,
    renderPlaceholder: PropTypes.func,
    renderLoading: PropTypes.func,
    renderError: PropTypes.func,
  };

  static defaultProps = {
    ratio: null,
    renderPlaceholder: () => null,
    renderLoading: () => <ProgressBar delay={0} />,
    renderError: () => <DefaultError />,
  };

  ref;

  observer;

  state = {
    state: 'not-in-view',
  };

  observer = getObserver();

  componentDidUpdate(prevProps) {
    const { src } = this.props;
    const { state } = this.state;

    if (prevProps.src !== src && state !== 'not-in-view') {
      this.setState(() => ({ state: 'loading' })); // eslint-disable-line
    }
  }

  componentWillUnmount() {
    this.teardownListeners();
  }

  setupListeners = () => {
    if (this.ref) {
      this.observer.observe(this.ref);
      emitter.on(events.IN_VIEW, this.handleIntersect);
    }
  };

  teardownListeners = () => {
    if (this.ref) {
      this.observer.unobserve(this.ref);
      emitter.off(events.IN_VIEW, this.handleIntersect);
    }
  };

  handleRef = ref => {
    if (ref) {
      this.ref = ref;
      this.setupListeners();
    }
  };

  handleIntersect = entry => {
    if (entry.target === this.ref) {
      this.setState({ state: 'loading' });
      this.teardownListeners();
    }
  };

  handleError = event => {
    const { onError } = this.props;
    this.setState(() => ({ state: 'error' }));
    if (onError) onError(event);
  };

  handleLoad = () => {
    this.setState(() => ({ state: 'loaded' }));
  };

  renderNotInView = () => {
    const { renderPlaceholder } = this.props;
    return renderPlaceholder();
  };

  renderLoading = () => {
    const { renderLoading, src } = this.props;
    return (
      <Fragment>
        {renderLoading()}
        <HiddenImage
          src={src}
          alt=""
          onLoad={this.handleLoad}
          onError={this.handleError}
          hidden
          aria-hidden
        />
      </Fragment>
    );
  };

  renderLoaded() {
    const { src, alt, className, onLoad } = this.props;
    return <Img src={src} alt={alt} className={className} onLoad={onLoad} />;
  }

  renderError = () => {
    const { renderError } = this.props;
    return renderError();
  };

  render() {
    const { ratio } = this.props;
    const { state } = this.state;

    return (
      <Container ratio={ratio} ref={this.handleRef}>
        {state === 'not-in-view' && this.renderNotInView()}
        {state === 'loading' && this.renderLoading()}
        {state === 'loaded' && this.renderLoaded()}
        {state === 'error' && this.renderError()}
      </Container>
    );
  }
}

export { LazyImage as default };
