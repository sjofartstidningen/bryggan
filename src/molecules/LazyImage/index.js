// @flow
import React, { PureComponent, Fragment } from 'react';
import type { Node } from 'react';
import { Container, HiddenImage, DefaultError } from './components';
import { Img } from '../../atoms/Image';
import getObserver, { emitter, events } from './observer';
import ProgressBar from '../ProgressBar';

type Props = {
  src: string,
  alt: string,
  ratio: number,
  className?: string,
  onLoad?: (event: SyntheticEvent<HTMLImageElement>) => void,
  onError?: (event: any) => void,
  renderPlaceholder: () => Node,
  renderLoading: () => Node,
  renderError: () => Node,
};

type State = {
  state: 'not-in-view' | 'loading' | 'loaded' | 'error',
};

class LazyImage extends PureComponent<Props, State> {
  static defaultProps = {
    ratio: 1,
    renderPlaceholder: () => null,
    renderLoading: () => <ProgressBar show delay={0} />,
    renderError: () => <DefaultError />,
  };

  ref: ?HTMLDivElement;
  observer: IntersectionObserver;

  state = {
    state: 'not-in-view',
  };

  observer = getObserver();

  componentDidUpdate(prevProps: Props) {
    const { src } = this.props;
    const { state } = this.state;

    if (prevProps.src !== src && state !== 'not-in-view') {
      this.setState(() => ({ state: 'loading' })); // eslint-disable-line
    }
  }

  componentWillUnmount() {
    this.teardownListeners();
  }

  setupListeners = () => {
    if (this.ref) {
      this.observer.observe(this.ref);
      emitter.on(events.IN_VIEW, this.handleIntersect);
    }
  };

  teardownListeners = () => {
    if (this.ref) {
      this.observer.unobserve(this.ref);
      emitter.off(events.IN_VIEW, this.handleIntersect);
    }
  };

  handleRef = (ref: ?HTMLDivElement) => {
    if (ref) {
      this.ref = ref;
      this.setupListeners();
    }
  };

  handleIntersect = (entry: IntersectionObserverEntry) => {
    if (entry.target === this.ref) {
      this.setState({ state: 'loading' });
      this.teardownListeners();
    }
  };

  handleError = (event: any) => {
    this.setState(() => ({ state: 'error' }));
    if (this.props.onError) this.props.onError(event);
  };

  handleLoad = () => {
    this.setState(() => ({ state: 'loaded' }));
  };

  renderNotInView = () => this.props.renderPlaceholder();

  renderLoading = () => (
    <Fragment>
      {this.props.renderLoading()}
      <HiddenImage
        src={this.props.src}
        alt=""
        onLoad={this.handleLoad}
        onError={this.handleError}
        hidden
        aria-hidden
      />
    </Fragment>
  );

  renderLoaded() {
    const { src, alt, className } = this.props;
    return (
      <Img
        src={src}
        alt={alt}
        className={className}
        onLoad={this.props.onLoad}
      />
    );
  }

  renderError = () => this.props.renderError();

  render() {
    const { ratio } = this.props;
    const { state } = this.state;

    return (
      <Container ratio={ratio} innerRef={this.handleRef}>
        {state === 'not-in-view' && this.renderNotInView()}
        {state === 'loading' && this.renderLoading()}
        {state === 'loaded' && this.renderLoaded()}
        {state === 'error' && this.renderError()}
      </Container>
    );
  }
}

export { LazyImage as default };
