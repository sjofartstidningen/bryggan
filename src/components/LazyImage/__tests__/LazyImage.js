/* eslint-disable no-underscore-dangle */
import React from 'react';
import { mount } from 'enzyme';
import axios from 'axios';
import LazyImage from '../';
import { setupIntersectionObserver, sleep } from '../../../test-utils';

jest.mock('axios');

describe('Component.LazyImage', () => {
  let reset;
  beforeAll(() => {
    reset = setupIntersectionObserver();

    window.URL.createObjectURL = () => 'src-string';

    axios.__setReturn({
      data: new Blob([JSON.stringify({ hello: 'world' }, null, 2)], {
        type: 'application/json',
      }),
    });
  });

  afterAll(() => {
    reset();
  });

  test('should download image once el is in view', async () => {
    const renderSpy = jest.fn(
      ({ src, loaded }) =>
        loaded ? <img className="inner" src={src} alt="" /> : null,
    );

    const wrapper = mount(
      <LazyImage src="path/to/image.jpg" render={renderSpy} />,
    );

    expect(renderSpy.mock.calls).toMatchSnapshot();

    const { observer } = wrapper.instance();
    observer.simulateIntersectionEvent({ isIntersecting: true });

    await sleep(0);
    expect(wrapper.state('src')).toBe('src-string');
    expect(renderSpy.mock.calls).toMatchSnapshot();
  });

  test('should reload image when calling reload from render-prop', async () => {
    const renderSpy = jest.fn(
      ({ loaded, reload }) =>
        loaded ? <button className="button" onClick={reload} /> : null,
    );

    const wrapper = mount(
      <LazyImage src="path/to/image.jpg" render={renderSpy} />,
    );

    const { observer } = wrapper.instance();
    observer.simulateIntersectionEvent({ isIntersecting: true });
    await sleep(0);
    wrapper.update();

    window.URL.createObjectURL = () => 'src-string-2';
    const button = wrapper.find('.button');
    button.simulate('click');

    await sleep(0);
    expect(wrapper.state('src')).toBe('src-string-2');
  });
});
