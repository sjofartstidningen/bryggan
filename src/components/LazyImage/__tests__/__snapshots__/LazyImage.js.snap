// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component.LazyImage should download image once el is in view 1`] = `
Array [
  Array [
    Object {
      "error": null,
      "loaded": false,
      "reload": [Function],
      "revokeObjectURL": [Function],
      "src": null,
    },
  ],
]
`;

exports[`Component.LazyImage should download image once el is in view 2`] = `
Array [
  Array [
    Object {
      "error": null,
      "loaded": false,
      "reload": [Function],
      "revokeObjectURL": [Function],
      "src": null,
    },
  ],
  Array [
    Object {
      "error": null,
      "loaded": true,
      "reload": [Function],
      "revokeObjectURL": [Function],
      "src": "src-string",
    },
  ],
]
`;
